bgpd/bgp_damp.c

```diff
@@ -37,6 +37,13 @@
 #include "bgpd/bgp_advertise.h"
 #include "bgpd/bgp_vty.h"
 
+static int reuselist_compare(const struct reuselist_node *o1,
+			     const struct reuselist_node *o2)
+{
+	return memcmp(&o1->info, &o2->info, sizeof(struct reuselist_node));
```

Comment:
This looks VERY incorrect.

First of all, you compare `struct bgp_damp_info` using `sizeof(struct reuselist_node)` so this is an overflow.
Second, there are a lot of moving parts in `struct bgp_damp_info` (`flap` is an example), so you can't just `memcmp` two structs to check that they are the same/different. I'm not familiar with the code, so I am not sure, but I suppose should use some combination of `path`, `dest`, `afi`, `safi` fields for comparison.

Reply:
Good catch, should be bgp_damp_info, it has already back pointers to path, dest, etc.

Reply:
It's not only about the size. You still can't compare the structs. As I said, there are a lot of changing fields there – flap, penalty, lastrecord, t_updated, maybe more. You have to find some fields that uniquely identify the RB entry and don't change over time to use them as a compare key.

Reply:
Will fix this.

Reply:
```
+static int bgp_damp_info_compare(const struct reuselist_node *o1,
+                          const struct reuselist_node *o2)
+{
+ struct bgp_damp_info *bdi1 = o1->info;
+ struct bgp_damp_info *bdi2 = o2->info;
+
+ if (bdi1->index < bdi2->index)
+         return -1;
+
+ if (bdi1->index > bdi2->index)
+         return 1;
+
+ if (bdi1->flap < bdi2->flap)
+         return -1;
+
+ if (bdi1->flap > bdi2->flap)
+         return 1;
+
+ if (bdi1->penalty < bdi2->penalty)
+         return -1;
+
+ if (bdi1->penalty > bdi2->penalty)
+         return 1;
+
+ if (bdi1->dest == bdi2->dest && bdi1->path == bdi2->path)
+         return 0;
+
+ return -1;
+}
```
@idryzhov is that fine for you?

Reply:
No, you still compare some changing fields – index, flap, penalty. You can't use any of these fields in the key, because when they change after the element is added to the tree, all your previous comparisons become invalid and your whole RB tree is essentially broken.

You can only use fields that don't ever change in the key for the RB tree. More than that, keys must be unique and comparable. So even if you remove all these invalid fields and keep only pointers comparison it is still not enough, because your tree will just become a simple list.

Thinking a little bit more about this – I don't think that using the RB tree here is a good idea. The whole point of the RB tree is to have a fast binary search, but there's not a single search in the dampening code. There are only inserts and removals. The best choice here would be to use double-linked lists.

Reply:
Nvm, closing.